sysl2/sysl/importer/xsd.go

```diff
@@ -0,0 +1,224 @@
+	var xsdToSyslMappings = map[string]string{
+		"NMTOKEN":      StringTypeName,
+		"integer":      "int",
+		"time":         StringTypeName,
+		"boolean":      "bool",
+		StringTypeName: StringTypeName,
+		"date":         "date",
+	}
+	for swaggerName, syslName := range xsdToSyslMappings {
+		from := xml.Name{Space: "http://www.w3.org/2001/XMLSchema", Local: swaggerName}
+		to := &SyslBuiltIn{syslName}
+		types.Add(makeNamespacedType(from, to))
+	}
+
+	keys := make([]xml.Name, 0, len(schema.Types))
+	for key := range schema.Types {
+		keys = append(keys, key)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return strings.Compare(
+			fmt.Sprintf("%s:%s", keys[i].Space, keys[i].Local),
+			fmt.Sprintf("%s:%s", keys[j].Space, keys[j].Local)) < 0
+	})
+
+	for _, name := range keys {
+		data := schema.Types[name]
+		logger.Infof("name: %+v\n", name)
+		if name.Local == "_self" {
+			rootType := data.(*xsd.ComplexType)
+			data = rootType.Elements[0].Type
+			t := FindType(data, &types)
+			if t == nil {
+				t = makeType(name, data, &types, logger)
+				if name.Space != "" {
+					types.Add(makeNamespacedType(name, t))
+				}
+			}
+			x := t.(*StandardType)
+			if !contains("~xml_root", x.Attributes) {
+				x.Attributes = append(x.Attributes, "~xml_root")
+			}
+		} else if t := FindType(data, &types); t == nil {
+			types.Add(makeType(name, data, &types, logger))
+		}
+	}
+
+	return types
+}
+
+func FindType(t xsd.Type, knownTypes *TypeList) Type {
+	if res, found := knownTypes.Find(fmt.Sprintf("%s:%s", xsd.XMLName(t).Space, xsd.XMLName(t).Local)); found {
+		return res
+	}
+	if res, found := knownTypes.Find(xsd.XMLName(t).Local); found {
+		return res
+	}
+	return nil
+}
+
+func makeType(_ xml.Name, from xsd.Type, knownTypes *TypeList, logger *logrus.Logger) Type {
+	switch t := from.(type) {
+	case *xsd.ComplexType:
+		return makeComplexType(t, knownTypes, logger)
+	case *xsd.SimpleType:
+		return makeSimpleType(t, knownTypes, logger)
+	case xsd.Builtin:
+		return makeXsdBuiltinType(t, knownTypes)
+	}
+	return nil
+}
+
+func makeComplexType(from *xsd.ComplexType, knownTypes *TypeList, logger *logrus.Logger) Type {
+	createChildItem := func(name xml.Name, data xsd.Type, isAttr, optional, plural bool) Field {
+		childType := FindType(data, knownTypes)
+		if childType == nil {
+			childType = makeType(name, data, knownTypes, logger)
+			knownTypes.Add(childType)
+		}
+		f := Field{
+			Name: name.Local,
+			Type: childType,
+		}
+		if isAttr {
+			f.Attributes = []string{"~xml_attribute"}
+		}
+		f.Optional = optional
+
+		if from, ok := data.(*xsd.SimpleType); ok && from.Restriction.Min > 0 {
+			spec := sizeSpec{
+				Min: int(from.Restriction.Min),
+				Max: int(from.Restriction.Max),
+			}
+			if spec.Max > 0 {
+				spec.MaxType = MaxSpecified
+			}
+			f.SizeSpec = &spec
+		}
+
+		if plural {
+			f.Type = &Array{Items: f.Type}
+		}
+		return f
+	}
+
+	item := &StandardType{
+		name: from.Name.Local,
+	}
+
+	for _, child := range from.Elements {
```

Comment:
Opportunity to trim duplicate code? The loops are alike.

Reply:
yes, if go had generics!